dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/NeuralNetwork.java

```diff
@@ -205,29 +210,29 @@ public void train(final INDArray input, final int label) {
                                         final INDArray[] activations,
                                         final INDArray nextError) {
     if (begin == layers.length - 1) {
-      throw new RuntimeException("The beginning layer of backPropagateFromTo cannot be output layer");
+      throw new IllegalArgumentException("The beginning layer of backPropagateFromTo cannot be the output layer");
     }
     if (begin < end) {
-      throw new RuntimeException(String.format(
+      throw new IllegalArgumentException(String.format(
           "The beginning index (%d) must be greater than or equal to the ending index (%d).", begin, end));
     }
 
-    checkIndices(begin, end, true);
+    checkIndices(begin, end, false);
 
-    final INDArray[] errors = new INDArray[end - begin + 1];
+    final INDArray[] errors = new INDArray[begin - end + 1];
     INDArray error = nextError;
 
     for (int i = begin; i >= end; --i) {
-      error = layers[i].backPropagate(activations[i + 1], layers[i + 1].getLayerParameter(), error);
-      errors[i - begin] = error;
+      error = layers[i].backPropagate(activations[i], activations[i + 1], error);
```

Comment:
In original implementation, A learnable layer (e.g. fully connected layer) calculates the errors that are used to generate gradients for training itself. For this computation, the next layer's parameter (weight and bias) is needed. 
This pull request changes that the errors used for generating gradients are calculated by the next layer. This makes the next layer's parameter unnecessary and learnable layers use their own parameters in backpropagation.